### ISSUE TITLE
rewire_edges: better handling of directed graphs

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1526]
`igraph_rewire_edges` effectively ignores edge directions, which means that directed graphs are not handled the way one would expect.

For example, if our graph is 1->2->3->1, i.e. a directed cycle graph, then it does not do any rewiring unless `multiple=true`. However, 1<->2->3 would be a valid _non-multigraph_ rewiring.

If our graph is 1->2, it never creates 2->1, even though this would be a valid directed rewiring.

This should be corrected.

We should think about whether it is useful to retain the old behaviour (i.e. simply ignore edge directions).

**Comment by GroteGnoom** [comment]
I want to work on this one, should I add a 'directed' parameter to ignore edge directions?

**Comment by szhorvat** [comment]
Please wait a bit with this, as this is part of a bigger project I have in mind. I need to recall the details and think it through again, but this week is super busy for me, so I might not get to it until Monday.

**Comment by szhorvat** [comment]
The first thing to check would be that the function works as it should in the undirected case. I have not done this yet, and there are some subtleties. Below I describe what I would like this function to do, but I do not know if that is indeed what it does.

This function first marks each edge with probability `prob`. Then it takes the marked edges, removes them, and re-inserts them in a random place in the graph. There are two cases: allow or disallow parallel edges. If we allow them, each edge is inserted between a randomly chosen pair of vertices. Here comes the first subtlety:

Each pair of vertex corresponds to one (i, j) location in the adjacency matrix. We should choose one such location _uniformly_ at random. This means, that with undirected / no-loops: choose a random location above the diagonal (below the diagonal will be its mirror image). With undirected / loops: choose a location on or above the diagonal uniformly. Make sure that a diagonal location is not chosen with a probability that differs from that of an above-diagonal location.

A naive implementation could e.g. choose both i and j at random between 1..n, then simply sort the pair (i,j), as we don't want to distinguish between i and j. This is wrong, as there is one way to choose (1,1), but two ways to choose (1,2), i.e. non-loops are created with a higher probability than loops. This is just a warning example: things can go wrong easily. Another wrong way to do it is to choose i between 1..n at random, then choose j between i..n at random. This is also not uniform.

The second case is when we disallow multi-edges. Suppose that there are `k` pairs of vertices in the graph (with different definition of "pair" for the undirected/directed and loop/no-loop cases) and that there are `m` edges of which we rewire `l`. Then `k - m + l \choose l` ways to re-insert the marked edges without creating multi-edges. The algorithm should be such that any of these configurations is created with the same probability.  I do not know if this is currently the case, but I was planning to verify that it is (and if not, to correct it).

**Comment by szhorvat** [comment]
There is a related task that I'm not sure if I created an issue for, and you may be interested in: `rewire_directed_edges` should support avoiding multi-edges. The same requirement applies: the sampling should be uniform. I.e., each possible outcome should appear with the same probability.

**Comment by szhorvat** [comment]
> should I add a 'directed' parameter to ignore edge directions?

I would say no. Directed graphs should be handled as directed, and undirected as undirected.

**Comment by ntamas** [comment]
The sampling of the upper triangle plus the diagonal for a matrix of size N x N can be done easily: pick a random integer _n_ between 0 and N * (N+1) / 2 (upper bound exclusive), then calculate the edge endpoints _u_ and _v_ as follows (Python variant, based on [this SO answer](https://stackoverflow.com/questions/40950460/how-to-convert-triangular-matrix-indexes-in-to-row-column-coordinates), but converted to zero-based indices):

```python
def conv(n):
    v = ceil(sqrt(2 * n + 1.25) - 1.5)
    u = n - v * (v + 1) // 2
    return u, v
```

I guess a similar formula exists also if the diagonal is not included.

**Comment by GroteGnoom** [comment]
> There is a related task that I'm not sure if I created an issue for, and you may be interested in: `rewire_directed_edges` should support avoiding multi-edges. The same requirement applies: the sampling should be uniform. I.e., each possible outcome should appear with the same probability.

Do you mean this one: https://github.com/igraph/igraph/issues/1527 ? I avoided that one because you are assigned to it. I'll fix #1527 if you want, :slightly_smiling_face:.

Something completely different: I think the process in igraph_rng_get_integer does not really get you a uniformly distributed integer in most cases. Is that an acceptable error? I think it's pretty hard to fix.

**Comment by szhorvat** [comment]
> I think the process in igraph_rng_get_integer does not really get you a uniformly distributed integer in most cases.

Why not?

The first branch does `(long int)(type->get_real(rng->state) * (h - l + 1) + l);`.  This implicitly assumes that `get_real` always returns numbers < 1, which may not always be the case and which I am not completely comfortable with. But the chance of failure (i.e. getting precisely 1) is very low.

The second branch does `(long int)(type->get(rng->state) / ((double)max + 1) * (h - l + 1) + l);`, which looks fine to me.

One thing I notice that we have a `min` value in `igraph_rng_type_t`, which is not being used. This is technically a bug. However, in practice all RNGs will have `min == 0`, so supporting this `min` value seems pointless to me ... I'd deprecate it.

**Comment by GroteGnoom** [comment]
> The second branch does `(long int)(type->get(rng->state) / ((double)max + 1) * (h - l + 1) + l);`, which looks fine to me.

This gets you all the right numbers, but they're not uniformly distributed.
For example if your origininal uniform distribution has a max of 2, and you want numbers between 0 and 1, then any calculation like this wil have an ouput of 3 numbers mapped to 2, so you'll get one number 1/3 of the time, and the other 2/3. You can fix this by taking a random number multiple times in some cases, but there will be no bound on the amount of times you have to do this.

I think it's the same issue as this one: https://rpg.stackexchange.com/a/175983

Of course the minimum maximum is 32767, so the problem is way smaller, but still there.

**Comment by szhorvat** [comment]
Of course, on a computer we must use discrete representations (a `float` and a `double` are also discrete), which leads to roundoff errors that might have significant effects in extreme cases. A usable RNG should have a large maximum value. The current generators in igraph have either 2^31 - 1 or 2^32 - 1 as the maximum, which should be sufficient in practice.

I guess one might come up with special crafted examples where it might matter, for example:

 - Suppose we have have a vector with 2/3 times the rand_max elements, and we want to choose an element at random. Every second element will be chosen with twice the probability.
 - Suppose that the vector elements are alternating between two values, e.g. 0, 1, 0, 1, 0, 1. Thus one of the values comes up with twice the probability. 

Ideally this would be improved at the source, i.e. the max element of the default generator should be much larger than the top random integer value we ever want to generate.

> Of course the minimum maximum is 32767

Do you mean the smallest maximum within igraph? This is currently 2^31 - 1.   There was a bug that the platform-dependent `RAND_MAX` was used for the glibc2 generator. I just corrected this.
